### PR TITLE
[move-vm][closures] Refactor: move abilities and closure mask into core types

### PR DESCRIPTION
**Pull request by wrwg**
## Description

[PR 2/n vm closures]

The type `AbilitySet` is required in `MoveTypeLayout`, and the type `ClosureMask` in `MoveValue`. This PRs moves those both types from file_format into the core types crate.

## How Has This Been Tested?

Existing tests

## Type of Change
- [ ] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [x] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [x] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Move Compiler
- [ ] Other (specify)

### CODE REVIEW

**Comment by wrwg** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/15668?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#15668** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/15668?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/15668?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#15667** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/15667?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: third_party/move/move-core/types/src/ability.rs [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+        I1::IntoIter: ExactSizeIterator,
+        I2::IntoIter: ExactSizeIterator,
+    {
+        let declared_phantom_parameters = declared_phantom_parameters.into_iter();
+        let type_arguments = type_arguments.into_iter();
+
+        if declared_phantom_parameters.len() != type_arguments.len() {
+            bail!("the length of `declared_phantom_parameters` doesn't match the length of `type_arguments`")
+            /*
+            return Err(
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-core/types/src/ability.rs:228]: This crate sits below move-binrary-format, so PartialVMError not available here.

We need this stuff here for MoveValue and MoveTypeLayout which is located here.

File: third_party/move/move-binary-format/src/binary_views.rs [hunk 2]
```diff
@@ -342,7 +343,11 @@ impl<'a> BinaryIndexedView<'a> {
             TypeParameter(idx) => Ok(constraints[*idx as usize]),
             Vector(ty) => AbilitySet::polymorphic_abilities(AbilitySet::VECTOR, vec![false], vec![
                 self.abilities(ty, constraints)?,
-            ]),
+            ])
+            .map_err(|e| {
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-binary-format/src/binary_views.rs:347]: Because polymorphic_abilities is now in the crate which does not have PartialVMError.

File: third_party/move/move-vm/types/src/loaded_data/runtime_types.rs [hunk 3]
```diff
@@ -625,6 +625,10 @@ impl Type {
                 AbilitySet::polymorphic_abilities(AbilitySet::VECTOR, vec![false], vec![
                     ty.abilities()?
                 ])
+                .map_err(|e| {
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-vm/types/src/loaded_data/runtime_types.rs:628]: No PartialVMError

File: third_party/move/move-core/types/src/function.rs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    pub fn bits(&self) -> u64 {
+        self.0
+    }
+
+    /// Apply a closure mask to a list of elements, returning only those
+    /// where position `i` is set in the mask (if `collect_captured` is true) or not
+    /// set (otherwise).
+    pub fn extract<'a, T>(&self, values: &'a [T], collect_captured: bool) -> Vec<&'a T> {
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-core/types/src/function.rs:41]: It's coming a bit upwards in the PR stack, IIRC.

File: third_party/move/move-core/types/src/function.rs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+pub struct ClosureMask(u64);
+
+impl fmt::Display for ClosureMask {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:b}", self.0)
+    }
+}
+
+impl ClosureMask {
+    pub const MAX_ARGS: usize = 64;
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-core/types/src/function.rs:28]: This is an interesting methodological disagreement.

I believe unit tests on this atomic level are counter productive. Better one can read the code and verify/"see" its working, as it is now the case. If the code would be so complex you need unit tests, something wrong with it. (As it was with the version in the previous PR.)

Indeed, the feature will be extensively tested by higher-level tests as it is a basic mechanism used in many places.

[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-core/types/src/function.rs:28]: PS. Bzgl. `MAX_ARGS`, I expect it used by the compiler. In the stack machine of the VM it plays no role because we never check "number of arguments correct", this is implicit via stack balancing.

File: third_party/move/move-core/types/src/ability.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply third_party/move/move-core/types/src/ability.rs:228]: I see, maybe add a comment about that? Looking at the code itself is not clear. At some point hopefully we can cleanup errors

File: third_party/move/move-core/types/src/function.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply third_party/move/move-core/types/src/function.rs:41]: Sounds good. In any case this doesn't matter to much for mvp

File: third_party/move/move-core/types/src/function.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply third_party/move/move-core/types/src/function.rs:28]: (1) Regarding unit tests:

I disagree. Unit test exists exactly for this purpose to verify small components work as expected and test corner cases. Good tests act as a spec by itself describing behaviour (there is a reason why TDD was created). Saying "my code works, just  read it and see" is not a great practice in my opinion: previous iteration of a mask had a panic on certain inputs - was it expected, unreachable code? These things should be documented and tested explicitly.

For example, since you derive `Arbitrary` and say that `mask.compose(mask.extract(v, true), mask.extract(v, false)) == v` should be preserved, a simple proptest doing this would be good. Actually, your `extract` right now skips elements > 64, so this property does not hold for all `v`.

[the remark this answers is not in the document]
Comment (georgemitenkov) [reply third_party/move/move-core/types/src/function.rs:28]: But let's also hear what other folks think: @ziaptos @vgao1996 @runtian-zhou @vineethk

**Review by georgemitenkov — APPROVED** [review]
I would strongly argue in favour of unit testing for the mask, but otherwise looks good

File: third_party/move/move-core/types/src/function.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-core/types/src/function.rs:33]: You want me to actually write a test for this code:

```
    let mut mask = self.0;
        values
            .iter()
            .filter(|_| {
                let set = mask & 0x1 != 0;
                mask >>= 1;
                set && collect_captured || !set && !collect_captured
            })
            .collect()
```

Sorry, I won't. The code is already the spec. 

Unnecessary unit tests are a kind of technical debt . You come back to change structure or semantics, the tests are a drag. This is a series of PRs where things evolve, and the basic concepts are revisited while solving each layer in the stack.

[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-core/types/src/function.rs:33]: > (2) Regarding number of arguments and MAX_ARGS: let's document t

Done

**Review by runtian-zhou — APPROVED** [review]

**Review by runtian-zhou — APPROVED** [review]
